### ISSUE TITLE
PR: Refine ggml-qnn backend(QNN, Qualcomm Neural Network,aka Qualcomm AI Engine Direct) for latest ggml,whisper.cpp,llama.cpp

### ISSUE THREAD

**Issue by zhouwg** [zhouwg/kantv#246]
----------------------------------------------------------------------------------------------------

|Required Info                         |                                                                |
|---------------------------------|------------------------------------------- |
| Branch                           |  { master }                          | 
| Device Vendor                 |   Xiaomi14(equipped with Qualcomm SM8650-AB Snapdragon 8 Gen 3)                    |
| Device OS Version          | Android 14                                |
| Section		|  { ggml-qnn }                          | 

### PR Description
there is a long story about this ggml-qnn backend, pls refer to:

 - my first touch(fall in love or "play") with ggml(03/05/2024---03/16/2024)  https://github.com/zhouwg/kantv/issues/64
 - my first implementation of ggml-qnn(03/29/2024---04/24/2024) https://github.com/zhouwg/kantv/issues/121

 - my first PR of ggml-qnn in upstream llama.cpp(04/24/2024---06/15/2024) https://github.com/ggerganov/llama.cpp/pull/6869

 - my refined implementation of ggml-qnn(01/29/2025---02/13/2025) https://github.com/zhouwg/kantv/issues/246

thanks to the big changes of software architecture(especially the "**backend scheduler**" feature has been introduced and matured) in latest upstream llama.cpp, this refined implementation can works pretty good with ASR inference via whisper.cpp and LLM inference via llama.cpp on Xiaomi14(equipped with Qualcomm Snapdragon 8 Gen 3).

[2025-02-10] **the source code of this PR are both available** in [project kantv](https://github.com/zhouwg/kantv/blob/master/core/ggml/llamacpp/ggml/src/ggml-qnn/ggml-qnn.cpp) and  [kantvai-ggmlqnn](https://github.com/kantv-ai/llama.cpp).
[2025-02-12] create <a href="https://github.com/kantv-ai">kantv.ai</a>, this PR will be submitted to the upstream llama.cpp from a **formal member** of [kantv-ai](https://github.com/kantv-ai) later after sanity-check and bug-fix again.
[2025-02-13] this PR has [submitted to the upstream llama.cpp community](https://github.com/ggerganov/llama.cpp/pull/11844) from a formal member of [the kantv-ai team](https://github.com/kantv-ai). any programmers interested in this topic pls drop your comment [in the upstream llama.cpp community](https://github.com/ggerganov/llama.cpp/pull/11844).

### How to verify the PR
 
this PR can be verified easily with a standard Android APP from master branch of project kantv(pls see below screenshots on 02-05-2024) or <b>the official test-backend-ops or llama-cli command line application</b> from [llama.cpp in kantv-ai](https://github.com/kantv-ai/llama.cpp).

- For Android developer, pls see [README-qnn.md in project kantv](https://github.com/zhouwg/kantv/blob/master/README-qnn.md)

- For llama.cpp community developer,

```
  git clone https://github.com/kantv-ai/llama.cpp
  cd llama.cpp
  git checkout kantvai-ggmlqnn
  ./build-run-android.sh build               (it'll setup local build envs automatically and build the entire project)
  ./build-run-android.sh updateqnnlib       (upload Qualcomm's QNN binary runtime libs to Android phone)
  ./build-run-android.sh run_llamacli        (running llama-cli on Android pohone)
  ./build-run-android.sh run_testop          (running test-backend-ops on Android phone)
```
you will find the "**aha moment**"  from the log output  of "adb logcat | grep KANTV".

### General notes

- **put everything in one single source file(ggml-qnn.cpp)**(because it's very **helpful** for other experienced programmers be involved  in dev activity which **similar to** what ggerganov did in the very beginning of ggml.c/llama.cpp or what Intel did in the very beginning of ggml-sycl.cpp) and enable it works pretty good  **before code reconstruction via C++**. if someone want to provide help with source codes or **participate in the dev activity of ggml-qnn**, pls **follow this coding style** and thanks for your corporation. 

- the previous and this refined implementation of ggml-qnn is mainly porting from <a href="https://github.com/pytorch/executorch">executorch</a>(the QNN backend's implementation in executorch comes from <b><a href="https://www.qualcomm.com/">Qualcomm</a></b>, especially got [**breakthrough** help from chiwwang@Qualcomm Technologies Inc](https://github.com/zhouwg/kantv/issues/121#issuecomment-2036126962). I also got a meaningful help from <a href="https://github.com/XiaoMi/StableDiffusionOnDevice">XiaoMi-StableDiffusionOnDevice</a>, so any other similar PRs in upstream llama.cpp are greatly welcomed so I can learning something from the PRs in upstream llama.cpp.

- of course,it's great that Qualcomm's QTI/QUIC can submit an official PR of ggml-qnn's implementation to the upstream llama.cpp which similar to what Intel/Huawei/AMD/Moore Threads did in the upstream llama.cpp.

### Additional notes

thanks for that I borrowed [ 5-7 functions ](https://github.com/zhouwg/kantv/blob/65cc05d52be2ae60ab5cac8b2fda8aa273843e33/core/ggml/llamacpp/ggml/src/ggml-qnn/ggml-qnn.cpp#L1536C1-L1590C2)  from a forked llama.cpp project which comes from a Chinese programmer <b>whom I don't know</b>. I'd like to cooperate with this Chinese programmer if he has intention of cooperate with me for ggml-qnn backend(**BUT** pls follow my coding style which I follow the coding style in the very beginning of ggml.c and <b>code reconstruction</b> via C++ is **NOT** the key point at current phase although I think the C++ skill of this unknown(<b>here</b> means "I don't know") Chinese programmer is good). I don't want to say anything others about this unknown(<b>here</b> means "I don't know") Chinese programmer because there is also a long story about that and the unrecoverable result has brought to me(I was blocked in the upstream llama.cpp community cause of his <b>unprofessional</b> behavior and my <b>stupid</b> comment in that PR in the upstream llama.cpp community), I don't want programmers from the US and EU to think this is a joke. 

<b>BTW</b>, my point of view about DeepSeek-R1: I'm not surprise that DeepSeek-R1 comes from China, because China has established the world's largest higher education system with <b>240 ------ 260 million</b> people having received university education and <b>accordingly</b> there are many STEM geniuses or <b>super smart</b> guys in China whom IQ is above 150 or similar to the original authors of the great llama.cpp. I agree that China's DeepSeek-R1 really brings a wake-up call to the US's AI&tech industry, but I <b>strongly</b> agree the point of view from Yann LeCun whom is Meta's Chief AI Scientist:

![Image](https://github.com/user-attachments/assets/ccd901f8-cff5-4489-9269-d6a46b1f1e7e)

I also <b>agree</b> the point of view from Oriol Vinyals whom is VP of Research & Deep Learning Lead from Google DeepMind:

![Image](https://github.com/user-attachments/assets/3faca4ce-084d-4e5b-91a2-929fcb628d56)

One of reasons of above point of review comes from an impressive sentence in a popular American song: "I won't forget the ones who died, who gave that right to me."

**Comment by zhouwg** [comment]
Thanks to the wonderful feature "backend scheduler" has been introduced in the latest upstream llama.cpp, we can implement a new ggml-qnn.cpp step by step in llama.cpp community.
The draft works fine as expected with whisper.cpp and llama.cpp on Xiaomi14.

**Comment by zhouwg** [comment]
**substantial or breakthrough** contributions/helps(DeepSeek-R1 compared to OpenAI o1) are greatly welcomed rather than code reconstruction at the C++ syntax level.

**Comment by zhouwg** [comment]
this draft is done on 02/09/2025.

**Comment by zhouwg** [comment]
any programmer whom interesting in the volunteer dev activity of ggml-qnn is greatly welcomed.pls file comments here.
